GitHubActionsTestLogger/TestReporterContext.cs

```diff
@@ -27,32 +29,42 @@ private string FormatAnnotation(string format, TestResult testResult)
 
         // Name token
         buffer
-            .Replace("@test", testResult.TestCase.DisplayName)
+            .Replace("@test", testNode.DisplayName)
             // Backwards compat
-            .Replace("$test", testResult.TestCase.DisplayName);
+            .Replace("$test", testNode.DisplayName);
 
         // Trait tokens
-        foreach (var trait in testResult.Traits.Union(testResult.TestCase.Traits))
+        // TODO: this is not enough, we would also need to handle properties from VSTest bridge
+        foreach (var metadataProperty in testNode.Properties.OfType<TestMetadataProperty>())
         {
             buffer
-                .Replace($"@traits.{trait.Name}", trait.Value)
+                .Replace($"@traits.{metadataProperty.Key}", metadataProperty.Value)
                 // Backwards compat
-                .Replace($"$traits.{trait.Name}", trait.Value);
+                .Replace($"$traits.{metadataProperty.Key}", metadataProperty.Value);
         }
 
+        Exception? exception = testNodeStateProperty switch
+        {
+            FailedTestNodeStateProperty failed => failed.Exception,
+            ErrorTestNodeStateProperty error => error.Exception,
+            _ => null,
+        };
+
         // Error message
         buffer
-            .Replace("@error", testResult.ErrorMessage ?? "")
+            .Replace("@error", exception?.Message ?? "")
             // Backwards compat
-            .Replace("$error", testResult.ErrorMessage ?? "");
+            .Replace("$error", exception?.Message ?? "");
 
         // Error trace
         buffer
-            .Replace("@trace", testResult.ErrorStackTrace ?? "")
+            .Replace("@trace", exception?.StackTrace ?? "")
             // Backwards compat
-            .Replace("$trace", testResult.ErrorStackTrace ?? "");
+            .Replace("$trace", exception?.StackTrace ?? "");
 
         // Target framework
+        // TODO: Copy logic from platform: https://github.com/microsoft/testfx/blob/main/src/Platform/Microsoft.Testing.Platform/OutputDevice/BrowserOutputDevice.cs#L78
```

Comment:
I guess as long as `TargetFrameworkParser` is public a one line copy-paste is fine?

Reply:
`Assembly.GetEntryAssembly()` would that not have any issues when running tests in an IDE? I guess with the MTP there is no special runner, so it shouldn't be a problem? Either way, I'd prefer not to take a logical dependency on that assumption if possible, so it would be great if the platform could expose the framework name.

Reply:
Assembly.GetEntryAssembly() is problem only for nativeaot and single-file mode, this to me looked like yet another case where we would benefit from telling the hook into which dll (or give it the hook type) they are hooking into.

https://github.com/microsoft/testfx/issues/5116